File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        
        

        int rtidx = 0; 
        return dfs(preorder,inorder , rtidx ,0 , inorder.size()-1) ; 
    }

    TreeNode* dfs(vector<int>& preorder , vector<int>& inorder, int& rtidx , int l , int r ){
    
        if(l > r) return NULL; 
        int pvt = l ; 
        while( inorder[pvt] != preorder[rtidx]) pvt ++ ; 

        rtidx ++; 
        TreeNode* nnode = new TreeNode(inorder[pvt]); 
        nnode->left = dfs(preorder,inorder , rtidx , l , pvt -1) ; 
        nnode->right = dfs(preorder, inorder, rtidx , pvt +1 ,r ) ; 
        return nnode ; 
}
        
        
        
    
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        
        if not preorder or not inorder: return 
        root = TreeNode(preorder[0])
        if len(preorder) ==1:  return root 
        
        rtidx = inorder.index(preorder[0])
        root.left = self.buildTree(preorder[1:rtidx+1] , inorder[:rtidx] )
        root.right = self.buildTree(preorder[rtidx+1:] , inorder[rtidx+1:])
        return root 